File: Sources/VoxLocalCore/Refinement/RefinementPipeline.swift
import Foundation

/// Applies optional refinement with all fallback guarantees: any provider
/// failure, timeout, cancellation of the provider call, or an implausible
/// result returns the raw transcript unchanged (with a local log warning).
public struct RefinementPipeline: Sendable {
    public struct Outcome: Equatable, Sendable {
        public let text: String
        /// True when the raw transcript was used (refinement skipped/failed).
        public let usedFallback: Bool
        public let fallbackReason: String?
    }

    private let provider: TextRefinementProvider

    public init(provider: TextRefinementProvider) {
        self.provider = provider
    }

    public func refine(_ transcript: String, context: RefinementContext) async -> Outcome {
        if context.preset == .rawTranscript {
            return Outcome(text: transcript, usedFallback: false, fallbackReason: nil)
        }
        do {
            let refined = try await provider.refine(transcript, context: context)
            switch RefinementSafeguard.validate(original: transcript, refined: refined) {
            case .accepted(let text):
                Log.shared.info("refinement accepted (\(transcript.count) → \(text.count) chars)")
                return Outcome(text: text, usedFallback: false, fallbackReason: nil)
            case .rejected(let reason):
                Log.shared.info("refinement rejected, using raw transcript: \(reason)")
                return Outcome(text: transcript, usedFallback: true, fallbackReason: reason)
            }
        } catch is CancellationError {
            Log.shared.info("refinement cancelled, using raw transcript")
            return Outcome(text: transcript, usedFallback: true, fallbackReason: "cancelled")
        } catch RefinementError.timeout {
            Log.shared.info("refinement timed out, using raw transcript")
            return Outcome(text: transcript, usedFallback: true, fallbackReason: "timeout")
        } catch {
            Log.shared.info("refinement failed, using raw transcript: \(error)")
            return Outcome(text: transcript, usedFallback: true, fallbackReason: String(describing: error))
        }
    }
}

File: windows/src/VoxLocal.App/Core/Refinement/RefinementPipeline.cs
using VoxLocal.Core.Utilities;

namespace VoxLocal.Core.Refinement;

/// <summary>
/// Applies optional refinement with all fallback guarantees: any provider
/// failure, timeout, cancellation of the provider call, or an implausible
/// result returns the raw transcript unchanged (with a local log warning).
/// </summary>
public sealed class RefinementPipeline
{
    /// <param name="UsedFallback">True when the raw transcript was used (refinement skipped/failed).</param>
    public sealed record Outcome(string Text, bool UsedFallback, string? FallbackReason);

    private readonly ITextRefinementProvider _provider;

    public RefinementPipeline(ITextRefinementProvider provider)
    {
        _provider = provider;
    }

    public async Task<Outcome> RefineAsync(
        string transcript,
        RefinementContext context,
        CancellationToken cancellationToken = default)
    {
        if (context.Preset == RefinementPreset.RawTranscript)
        {
            return new Outcome(transcript, UsedFallback: false, FallbackReason: null);
        }
        try
        {
            var refined = await _provider.RefineAsync(transcript, context, cancellationToken).ConfigureAwait(false);
            switch (RefinementSafeguard.Validate(transcript, refined))
            {
                case RefinementSafeguard.Verdict.Accepted accepted:
                    Log.Shared.Info($"refinement accepted ({transcript.Length} -> {accepted.Text.Length} chars)");
                    return new Outcome(accepted.Text, UsedFallback: false, FallbackReason: null);
                case RefinementSafeguard.Verdict.Rejected rejected:
                    Log.Shared.Info($"refinement rejected, using raw transcript: {rejected.Reason}");
                    return new Outcome(transcript, UsedFallback: true, rejected.Reason);
                default:
                    return new Outcome(transcript, UsedFallback: true, "unknown verdict");
            }
        }
        catch (OperationCanceledException)
        {
            Log.Shared.Info("refinement cancelled, using raw transcript");
            return new Outcome(transcript, UsedFallback: true, "cancelled");
        }
        catch (RefinementTimeoutException)
        {
            Log.Shared.Info("refinement timed out, using raw transcript");
            return new Outcome(transcript, UsedFallback: true, "timeout");
        }
        catch (Exception e)
        {
            Log.Shared.Info($"refinement failed, using raw transcript: {e.Message}");
            return new Outcome(transcript, UsedFallback: true, e.ToString());
        }
    }
}
